File: SubarraySumEqualsK.java
/**
 * https://leetcode.com/problems/subarray-sum-equals-k/
 * @date 2020-08-05
 */
public class SubarraySumEqualsK {
    public static void main(String[] args) {
        int nums[] = {0, 0, 0, 0};
        int k = 0;
        assert subarraySum(nums, k) == 10;
    }

    private static int subarraySum(int[] nums, int k) {
        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            int sum = nums[i];
            boolean flag = false;
            if (sum == k) {
                count++;
                flag = true;
            }
            for (int j = i + 1; j < nums.length; j++) {
                sum += nums[j];
                if (sum == k) {
                    count++;
                    flag = true;
                }
            }
            if (!flag && sum == k) {
                count++;
            }
        }
        return count;
    }
}

File: rust/subarray_sum_equals_k.rs
//! <https://leetcode.com/problems/subarray-sum-equals-k/>
use std::collections::HashMap;
pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
    let mut map = HashMap::from([(0, 1)]);
    let mut sum = 0;
    let mut count = 0;
    for num in nums {
        sum += num;
        count += map.get(&(sum - k)).unwrap_or(&0);
        map.entry(sum).and_modify(|v| *v += 1).or_insert(1);
    }
    count
}

fn main() {
    assert_eq!(subarray_sum(vec![1, 1, 1], 2), 2);
    assert_eq!(subarray_sum(vec![1, 2, 3], 3), 2);
    assert_eq!(subarray_sum(vec![1, -1, 1, 1, 1, 1], 3), 4);
}
